hibernate-core/src/main/java/org/hibernate/type/CollectionType.java

```diff
@@ -712,7 +712,12 @@ public Object replace(
 			final Object owner,
 			final Map copyCache) throws HibernateException {
 		if ( original == null ) {
-			return null;
+			if(target instanceof PersistentCollection) {
+				return target;
+			}
+			else {
+				return null;
+			}
```

Comment:
Your first statement in this comment is incorrect, not only because this PR is passing almost all test cases in Hibernate Core, but also because cascade processing has already occurred by this time.  

In the above method, if `target` is a `PersistentCollection` and the original value was `null`, we are currently processing a lazy loaded collection property that Hibernate Search may need to traverse.  We need to assign the uninitialized proxy value to the object.  The purpose of this code is to not overlay an uninitialized proxy collection with the originally null value, so Hibernate Search can traverse the relationship, and possibly go deeper into the hierarchical tree by traversing even beyond properties of the present entity instance.

Reply:
It is an incorrect assumption that `null` always means empty, even in ORM today, and this ambiguity is the root cause of our problems.  ORM needs to change this to resolve HHH-14839 with a long term solution.  As discussed previously in the forum, this PR is a workaround pending a fundamental architecture change to ORM.

ORM is inconsistent in that it NEVER returns a null value for a lazy loaded collection property after loading an entity, but always returns null collection values passed in to `em.persist` and `em.merge`, even when they should have data.  

**The main effect of this PR is to standardize results returned from `em.persist` and `em.merge` so that they are more consistent with those of `em.find`.**

ORM cannot carry forward bad design forever.  At some point it must be realized that a breaking change is absolutely necessary to purge bad design.  Long term, if a null collection will not mean uninitialized, then there needs to be a placeholder for collections that actually means uninitialized in all cases.

It is our opinion that the meaning of `null` must be changed from a concept of "empty" to "undefined" **_when assigned to a collection property_**, and that only a collection object with length zero should be seen as an empty collection.

Note also that at the present time, the code in this PR does not affect associations which cascade in any way, to limit the complexity initially.  It might be beneficial to Hibernate Search to have all cascading collections transformed into an uninitialized proxy as well, but we hardly ever use them in our application, so we did not address this yet.

If we _were_ to apply this logic to collections that use cascade features, applications would simply see a more valid way of representing an empty collection returned instead of null.  Applications would see an actual collection object with length of zero, instead of what was previously `null` and ambiguous.

If such an application were depending on seeing null instead of a collection with zero length, it would have to change slightly, but as I already stated, ORM _already_ provides applications with such a value for ALL lazy collection properties during `em.find`, so they are already dealing with them there...